File: Scripts/validate_lua.py
#!/usr/bin/env python3
"""
Validate Lua syntax for Gargul addon files.

Uses luac -p (Lua 5.1 compiler) when available. WoW uses Lua 5.1, so luac
provides authoritative syntax validation.

Usage:
  python Scripts/validate_lua.py                    # validate all .lua files
  python Scripts/validate_lua.py path/to/file.lua  # validate specific file(s)
  python Scripts/validate_lua.py Classes/          # validate directory
  python Scripts/validate_lua.py --luac "C:\\path\\to\\luac5.1.exe"  # if not in PATH

Install Lua 5.1 (for luac). WoW uses Lua 5.1; winget's Lua is 5.4.
  - Windows: LuaBinaries 5.1.5 from https://sourceforge.net/projects/luabinaries/files/5.1.5/Tools%20Executables/
  - Or: https://www.lua.org/download.html
  - Add folder to PATH (LuaBinaries provides luac5.1.exe)
"""

import argparse
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import List, Optional, Tuple

UTF8_BOM = b"\xef\xbb\xbf"


# Directories to skip (Libs are third-party; Data may have non-standard Lua)
SKIP_DIRS = {".git", "Libs", "automations", "Data", "Scripts"}


def find_lua_files(root: Path, paths: Optional[List[Path]]) -> List[Path]:
    """Collect .lua files to validate."""
    if paths:
        files = []
        for p in paths:
            p = Path(p)
            if p.is_file() and p.suffix == ".lua":
                files.append(p)
            elif p.is_dir():
                files.extend(f for f in p.rglob("*.lua") if not any(s in f.parts for s in SKIP_DIRS))
        return sorted(set(files))

    # Default: all .lua in project, excluding skip dirs
    root = Path(root)
    return [
        f
        for f in root.rglob("*.lua")
        if not any(s in f.parts for s in SKIP_DIRS)
    ]


def validate_with_luac(file_path: Path, luac_cmd: Optional[str] = None) -> Tuple[bool, str]:
    """Validate using luac -p. Returns (ok, error_message)."""
    # Lua 5.1 chokes on UTF-8 BOM; strip it to a temp file for validation
    raw = file_path.read_bytes()
    if raw.startswith(UTF8_BOM):
        raw = raw[len(UTF8_BOM) :]
    with tempfile.NamedTemporaryFile(suffix=".lua", delete=False) as tmp:
        tmp.write(raw)
        tmp_path = tmp.name
    try:
        cmds = [luac_cmd] if luac_cmd else ["luac5.1", "luac"]
        for cmd in cmds:
            if not cmd:
                continue
            try:
                result = subprocess.run(
                    [cmd, "-p", tmp_path],
                    capture_output=True,
                    text=True,
                    timeout=5,
                )
                if result.returncode == 0:
                    return True, ""
                return False, (result.stderr or result.stdout or "unknown error").strip()
            except FileNotFoundError:
                continue
            except subprocess.TimeoutExpired:
                return False, "timeout"
        return False, "luac not found (try --luac path or add luac5.1 to PATH)"
    finally:
        try:
            Path(tmp_path).unlink()
        except OSError:
            pass


def main():
    parser = argparse.ArgumentParser(description="Validate Lua syntax for Gargul addon files")
    parser.add_argument(
        "paths",
        nargs="*",
        help="Files or directories to validate (default: all .lua in project)",
    )
    parser.add_argument(
        "--root",
        default=Path(__file__).resolve().parent.parent,
        type=Path,
        help="Project root (default: parent of Scripts/)",
    )
    parser.add_argument(
        "--luac",
        metavar="PATH",
        help="Path to luac5.1.exe (if not in PATH)",
    )
    args = parser.parse_args()

    paths = [Path(p) for p in args.paths] if args.paths else None
    files = find_lua_files(args.root, paths)
    if not files:
        print("No .lua files found to validate.")
        sys.exit(0)

    errors = []
    for f in files:
        ok, msg = validate_with_luac(f, args.luac)
        try:
            rel = f.relative_to(args.root)
        except ValueError:
            rel = f
        if ok:
            print(f"OK   {rel}")
        else:
            print(f"FAIL {rel}")
            if msg:
                for line in msg.splitlines():
                    print(f"     {line}")
            errors.append((rel, msg))

    if errors:
        print(f"\n{len(errors)} file(s) failed validation.")
        if errors and "luac not found" in errors[0][1]:
            print("\nInstall Lua 5.1 to enable validation (WoW uses 5.1; winget's Lua is 5.4):")
            print("  LuaBinaries: https://sourceforge.net/projects/luabinaries/files/5.1.5/Tools%20Executables/")
            print("  Or: https://www.lua.org/download.html")
            print("  Use --luac path/to/luac5.1.exe or add luac5.1 to PATH.")
        sys.exit(1)

    print(f"\nAll {len(files)} file(s) passed.")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: Scripts/validate_lua.lua
--[[
    Validate Lua 5.1 syntax for Gargul addon files, using the interpreter's own parser.

    Run it with the bundled interpreter (no install needed):
      Scripts\bin\lua5.1.exe Scripts\validate_lua.lua                      -- every .lua file
      Scripts\bin\lua5.1.exe Scripts\validate_lua.lua Classes\Player.lua   -- specific files
      Scripts\bin\lua5.1.exe Scripts\validate_lua.lua Classes              -- a directory
]]

local SKIP_DIRS = { ".git", "Libs", "automations", "Data", "Scripts", };

---@return string
local function currentDirectory()
    local Pipe = io.popen("cd");
    local path = Pipe:read("*l");
    Pipe:close();

    return path;
end

--- Turn a path into a backslashed absolute-ish form with "." and ".." resolved.
---@return string
local function normalize(path)
    path = path:gsub("/", "\\");
    path = path:gsub("\\+$", "");
    path = path:gsub("\\%.\\", "\\");
    path = path:gsub("\\%.$", "");

    while (true) do
        local resolved, count = path:gsub("\\[^\\]+\\%.%.", "", 1);
        if (count == 0) then
            break;
        end

        path = resolved;
    end

    return path;
end

---@return string
local function absolute(path)
    path = normalize(path);
    if (path:match("^%a:")) then
        return path;
    end

    return normalize(currentDirectory() .. "\\" .. path);
end

local scriptPath = normalize(arg[0]);
local root = absolute((scriptPath:match("^(.*)\\[^\\]*$") or ".") .. "\\..");

---@return string
local function relativePath(path)
    if (path:sub(1, #root + 1):lower() == (root .. "\\"):lower()) then
        return path:sub(#root + 2);
    end

    return path;
end

---@return boolean
local function skipped(name)
    local padded = "\\" .. name .. "\\";
    for _, directory in ipairs(SKIP_DIRS) do
        if (padded:find("\\" .. directory .. "\\", 1, true)) then
            return true;
        end
    end

    return false;
end

local Files = {};
local seen = {};

---@return nil
local function addFile(path)
    path = absolute(path);

    local name = relativePath(path);
    if (seen[name:lower()]) then
        return;
    end

    seen[name:lower()] = true;
    Files[#Files + 1] = { path = path, name = name, };
end

--- Add every .lua file below a directory, minus the ones we never check.
---@return nil
local function addDirectory(directory)
    local Pipe = io.popen(("dir /b /s \"%s\\*.lua\" 2>nul"):format(normalize(directory)));
    for line in Pipe:lines() do
        local path = normalize(line);
        if (not skipped(relativePath(path))) then
            addFile(path);
        end
    end
    Pipe:close();
end

--- Compile the file without running it. Returns false plus the parse error on bad syntax.
---@return boolean, string?
local function validate(File)
    local Handle, openError = io.open(File.path, "rb");
    if (not Handle) then
        return false, openError or "could not open file";
    end

    local source = Handle:read("*a");
    Handle:close();

    -- Lua 5.1 refuses to parse a UTF-8 BOM, WoW strips it for us.
    source = source:gsub("^\239\187\191", "");

    local chunk, syntaxError = loadstring(source, "@" .. File.name);
    if (chunk) then
        return true;
    end

    return false, syntaxError;
end

if (#arg == 0) then
    addDirectory(root);
else
    for _, target in ipairs(arg) do
        if (target:lower():sub(-4) == ".lua") then
            addFile(target);
        else
            addDirectory(target);
        end
    end
end

if (#Files == 0) then
    print("No .lua files found to validate.");
    os.exit(0);
end

table.sort(Files, function (First, Second)
    return First.name:lower() < Second.name:lower();
end);

local failures = 0;
for _, File in ipairs(Files) do
    local ok, message = validate(File);

    if (ok) then
        print("OK   " .. File.name);
    else
        failures = failures + 1;
        print("FAIL " .. File.name);
        print("     " .. tostring(message));
    end
end

print("");
if (failures > 0) then
    print(("%d of %d file(s) failed validation."):format(failures, #Files));
    os.exit(1);
end

print(("All %d file(s) passed."):format(#Files));
os.exit(0);
